### ISSUE TITLE
[JENKINS-54956] Github merge refs not working correctly

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/github-branch-source-plugin#1225]
<p>We're having some issues where some of our developer's pull requests are not reporting that the "Jenkinsfile not found" when we know that their branch does indeed have the Jenkinsfile.</p>

<p>We first noticed it with GitHub webhooks that trigger the event when there is a change to the pull request.. that also reports "Jenkinsfile not found". Here is a snippet of the log: <br/>
``` <br/>
<span class="error">&#91;Fri Nov 23 09:13:55 PST 2018&#93;</span> Pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35771" class="external-link" target="_blank" rel="nofollow noopener">#35771</a> updated in repository Safe/fme UPDATED event from xxx.xxx.xxx.xx → xxx.xxx.xxx.xx → xxx.xxx.xxx.xx⇒ <a href="http://cje-1.base.safe.com/teams-fme-engine/github-webhook/" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxx/github-webhook/</a> with timestamp Fri Nov 23 09:13:48 PST 2018 processed in 2 sec <br/>
<span class="error">&#91;Fri Nov 23 09:27:55 PST 2018&#93;</span> Received Pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a> opened in repository Safe/fme CREATED event from xxx.xxx.xx.xxx → xxx.xxx.xxx.xxx → xxx.xxx.xxx.xx ⇒ <a href="http://cje-1.base.safe.com/teams-fme-engine/github-webhook/" class="external-link" target="_blank" rel="nofollow noopener">xxxxxxxxx/github-webhook/</a> with timestamp Fri Nov 23 09:27:50 PST 2018<br/>
09:27:56 Connecting to <a href="http://gtb-1.base.safe.com/api/v3" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxx/api/v3</a> using bbuilder/****** (Bob the Builder's Credentials) <br/>
Examining Safe/fme</p>

<p>Checking pull-requests...</p>

<p>Getting remote pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a>...</p>

<p>Checking pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a> <br/>
‘jenkins/Jenkinsfile’ not found <br/>
Does not meet criteria</p>

<p>Checking pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a> <br/>
‘jenkins/Jenkinsfile’ not found <br/>
Does not meet criteria</p>

<p>2 pull requests were processed</p>

<p>Finished examining Safe/fme <br/>
```</p>

<p>We then tried to use the "Scan Repository Now" feature to see if it would find it then.. but that also failed.</p>

<p>We dug into the plugin source code and found that it was trying to access this API endpoint on GitHub:</p>

<p><a href="http://gtb-1.base.safe.com/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxxx/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge</a></p>

<p>The response from that API call was this:</p>

<p>``` <br/>
{ <br/>
"message": "No commit found for the ref pull/35897/merge", <br/>
"documentation_url": "<a href="https://developer.github.com/enterprise/2.14/v3/repos/contents/" class="external-link" target="_blank" rel="nofollow noopener">https://developer.github.com/enterprise/2.14/v3/repos/contents/</a>" <br/>
} <br/>
```</p>

<p>After some time, the API somehow refreshed and it now properly returns the right contents and the pipeline is now working for that pull request.</p>

<p> </p>

<p>Aparently this API call to get the '/merge' refs is an undocumented feature and not fully supported by GitHub.</p>

<p><a href="http://gtb-1.base.safe.com/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxxxx/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge</a></p>

<p>Here is their response to that:</p>

<p>---------------------------------- <br/>
Hi xxxx,</p>

<p>The /merge refs are an undocumented feature and we don't recommend relying on them as they're subject to change at anytime. This feature was built to power the Pull Requests feature in the GitHub Web UI, it was not built to be used by 3rd party services or tooling such as an API.</p>

<p>As this is undocumented, you should not have any expectations about the behavior, the behavior might change at any time and those refs might completely go away without warning. In our current implementation, the pull/*/merge reference should be eventually updated when either the head or base branch of the pull request being pushed to. This happens in the background that is run asynchronously and GitHub Enterprise does not guarantee that it will be available at any specific time after a pull request's head branch is pushed to.</p>

<p>It's also possible for the merge ref to be missing or point to an out-of-date merge; for example, if the latest push to a branch causes a merge conflict.</p>

<p>Again, we do not recommend relying on this ref and instead recommend you update your jobs to use the head branch. For example:</p>

<p><a href="http://gtb-1.base.safe.com/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/head" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxx/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/head</a> <br/>
This should always be available after a push completes. If you'd like to test the result of a merge with the base branch, your CI server can either perform the merge locally, or you can <span class="error">&#91;configure a required status check with the strict policy&#93;</span>(<a href="https://help.github.com/enterprise/2.14/user/articles/types-of-required-status-checks/" class="external-link" target="_blank" rel="nofollow noopener">https://help.github.com/enterprise/2.14/user/articles/types-of-required-status-checks/</a>, which will ensure that the head branch is always up-to-date with respect to the base before merging is allowed.</p>

<p>Let us know how that works for you and please do not hesitate to reach out with any further questions.</p>

<p>Thanks,</p>

<p>Brian <br/>
----------------------------------</p>

<p>With that, it seems the GitHub Branch Source Plugin is may not be very reliable and could potentially break at anytime. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jraezrus">jraezrus</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54956" target="_blank">Github merge refs not working correctly</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: github-branch-source-plugin
<li><b>resolution</b>: Cannot Reproduce
<li><b>resolved</b>: 2019-05-29T18:35:18+00:00
<li><b>votes</b>: 1
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We're having some issues where some of our developer's pull requests are not reporting that the "Jenkinsfile not found" when we know that their branch does indeed have the Jenkinsfile.</p>

<p>We first noticed it with GitHub webhooks that trigger the event when there is a change to the pull request.. that also reports "Jenkinsfile not found". Here is a snippet of the log: 
``` 
<span class="error">&#91;Fri Nov 23 09:13:55 PST 2018&#93;</span> Pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35771" class="external-link" target="_blank" rel="nofollow noopener">#35771</a> updated in repository Safe/fme UPDATED event from xxx.xxx.xxx.xx → xxx.xxx.xxx.xx → xxx.xxx.xxx.xx⇒ <a href="http://cje-1.base.safe.com/teams-fme-engine/github-webhook/" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxx/github-webhook/</a> with timestamp Fri Nov 23 09:13:48 PST 2018 processed in 2 sec 
<span class="error">&#91;Fri Nov 23 09:27:55 PST 2018&#93;</span> Received Pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a> opened in repository Safe/fme CREATED event from xxx.xxx.xx.xxx → xxx.xxx.xxx.xxx → xxx.xxx.xxx.xx ⇒ <a href="http://cje-1.base.safe.com/teams-fme-engine/github-webhook/" class="external-link" target="_blank" rel="nofollow noopener">xxxxxxxxx/github-webhook/</a> with timestamp Fri Nov 23 09:27:50 PST 2018
09:27:56 Connecting to <a href="http://gtb-1.base.safe.com/api/v3" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxx/api/v3</a> using bbuilder/****** (Bob the Builder's Credentials) 
Examining Safe/fme</p>

<p>Checking pull-requests...</p>

<p>Getting remote pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a>...</p>

<p>Checking pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a> 
‘jenkins/Jenkinsfile’ not found 
Does not meet criteria</p>

<p>Checking pull request <a href="https://cloudbees.zendesk.com/agent/tickets/35897" class="external-link" target="_blank" rel="nofollow noopener">#35897</a> 
‘jenkins/Jenkinsfile’ not found 
Does not meet criteria</p>

<p>2 pull requests were processed</p>

<p>Finished examining Safe/fme 
```</p>

<p>We then tried to use the "Scan Repository Now" feature to see if it would find it then.. but that also failed.</p>

<p>We dug into the plugin source code and found that it was trying to access this API endpoint on GitHub:</p>

<p><a href="http://gtb-1.base.safe.com/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxxx/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge</a></p>

<p>The response from that API call was this:</p>

<p>``` 
{ 
"message": "No commit found for the ref pull/35897/merge", 
"documentation_url": "<a href="https://developer.github.com/enterprise/2.14/v3/repos/contents/" class="external-link" target="_blank" rel="nofollow noopener">https://developer.github.com/enterprise/2.14/v3/repos/contents/</a>" 
} 
```</p>

<p>After some time, the API somehow refreshed and it now properly returns the right contents and the pipeline is now working for that pull request.</p>

<p> </p>

<p>Aparently this API call to get the '/merge' refs is an undocumented feature and not fully supported by GitHub.</p>

<p><a href="http://gtb-1.base.safe.com/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxxxx/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/merge</a></p>

<p>Here is their response to that:</p>

<p>---------------------------------- 
Hi xxxx,</p>

<p>The /merge refs are an undocumented feature and we don't recommend relying on them as they're subject to change at anytime. This feature was built to power the Pull Requests feature in the GitHub Web UI, it was not built to be used by 3rd party services or tooling such as an API.</p>

<p>As this is undocumented, you should not have any expectations about the behavior, the behavior might change at any time and those refs might completely go away without warning. In our current implementation, the pull/*/merge reference should be eventually updated when either the head or base branch of the pull request being pushed to. This happens in the background that is run asynchronously and GitHub Enterprise does not guarantee that it will be available at any specific time after a pull request's head branch is pushed to.</p>

<p>It's also possible for the merge ref to be missing or point to an out-of-date merge; for example, if the latest push to a branch causes a merge conflict.</p>

<p>Again, we do not recommend relying on this ref and instead recommend you update your jobs to use the head branch. For example:</p>

<p><a href="http://gtb-1.base.safe.com/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/head" class="external-link" target="_blank" rel="nofollow noopener">http://xxxxxx/api/v3/repos/Safe/fme/contents/jenkins?ref=pull/35897/head</a> 
This should always be available after a push completes. If you'd like to test the result of a merge with the base branch, your CI server can either perform the merge locally, or you can <span class="error">&#91;configure a required status check with the strict policy&#93;</span>(<a href="https://help.github.com/enterprise/2.14/user/articles/types-of-required-status-checks/" class="external-link" target="_blank" rel="nofollow noopener">https://help.github.com/enterprise/2.14/user/articles/types-of-required-status-checks/</a>, which will ensure that the head branch is always up-to-date with respect to the base before merging is allowed.</p>

<p>Let us know how that works for you and please do not hesitate to reach out with any further questions.</p>

<p>Thanks,</p>

<p>Brian 
----------------------------------</p>

<p>With that, it seems the GitHub Branch Source Plugin is may not be very reliable and could potentially break at anytime. </p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>thomas_deepomatic:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54956?focusedId=359087&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-359087">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We encountered the same issue. After 5h on github.com the `/merge` ref is still unavailable it seems...</p>

<p>The only workaround was to create a new PR.</p>

<p>I tested:</p>
<ul class="alternate" type="square">
	<li>closing/reopening</li>
	<li>pushing new commits</li>
	<li>waiting 4h</li>
</ul>


<p>According to github it should eventually converge, maybe we should loop wait on jenkins side? But 4h is long, maybe there is a deadlock on github side when on github the branch is protected by waiting for jenkins CI status?</p>

<p>Suggestion: at least the jenkins plugin should log something different than `‘Jenkinsfile’ not found`. Maybe an explicit explanation linking to here, and document the workaround: create a new PR.</p>

</pre>
</details>

</ul>
</details>

<p>We encountered the same issue. After 5h on github.com the `/merge` ref is still unavailable it seems...</p>

<p>The only workaround was to create a new PR.</p>

<p>I tested:</p>
<ul class="alternate" type="square">
	<li>closing/reopening</li>
	<li>pushing new commits</li>
	<li>waiting 4h</li>
</ul>


<p>According to github it should eventually converge, maybe we should loop wait on jenkins side? But 4h is long, maybe there is a deadlock on github side when on github the branch is protected by waiting for jenkins CI status?</p>

<p>Suggestion: at least the jenkins plugin should log something different than `‘Jenkinsfile’ not found`. Maybe an explicit explanation linking to here, and document the workaround: create a new PR.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bitwiseman.png" title="bitwiseman's avatar" /> bitwiseman:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54956?focusedId=366181&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-366181">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_deepomatic" class="user-hover" rel="thomas_deepomatic">thomas_deepomatic</a>, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jraezrus" class="user-hover" rel="jraezrus">jraezrus</a>
4h is particularly long. </p>

<p>That email if very confusing. I guess it is technically true, 'refs/pulls/nnnn/merge' is not specifically documented, but multiple things related to it are.
<tt>Merge_commit_sha</tt> and <tt>refs/pulls/nnnn/head</tt> are directly connected to this and have documentation. 
<a href="https://developer.github.com/v3/pulls/#get-a-single-pull-request" class="external-link" target="_blank" rel="nofollow noopener">https://developer.github.com/v3/pulls/#get-a-single-pull-request</a> (see response). </p>

<p>Could you try out github branch source plugin v2.5.1. This may address this issue for you - it works more deeply but also in a more accurate way with these APIs.  If nothing else, it would be very useful to have more data from  the current version of the plugin. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_deepomatic" class="user-hover" rel="thomas_deepomatic">thomas_deepomatic</a>, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jraezrus" class="user-hover" rel="jraezrus">jraezrus</a><br/>
4h is particularly long. </p>

<p>That email if very confusing. I guess it is technically true, 'refs/pulls/nnnn/merge' is not specifically documented, but multiple things related to it are.<br/>
<tt>Merge_commit_sha</tt> and <tt>refs/pulls/nnnn/head</tt> are directly connected to this and have documentation. <br/>
<a href="https://developer.github.com/v3/pulls/#get-a-single-pull-request" class="external-link" target="_blank" rel="nofollow noopener">https://developer.github.com/v3/pulls/#get-a-single-pull-request</a> (see response). </p>

<p>Could you try out github branch source plugin v2.5.1. This may address this issue for you - it works more deeply but also in a more accurate way with these APIs.  If nothing else, it would be very useful to have more data from  the current version of the plugin. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bitwiseman.png" title="bitwiseman's avatar" /> bitwiseman:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-54956?focusedId=368087&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-368087">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jraezrus" class="user-hover" rel="jraezrus">jraezrus</a><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_deepomatic" class="user-hover" rel="thomas_deepomatic">thomas_deepomatic</a>
Could you try out github branch source plugin v2.5.3? </p>

<p>I'm going to resolve this this due to lack of response.  We can reopen if it still occurring. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jraezrus" class="user-hover" rel="jraezrus">jraezrus</a><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=thomas_deepomatic" class="user-hover" rel="thomas_deepomatic">thomas_deepomatic</a><br/>
Could you try out github branch source plugin v2.5.3? </p>

<p>I'm going to resolve this this due to lack of response.  We can reopen if it still occurring. </p>
